File: CodingTest/Java/Solution22.java
// https://school.programmers.co.kr/learn/courses/30/lessons/131127
package CodingTest.Java;

import java.util.HashMap;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.Map;
import java.util.Queue;

public class Solution22 {
    public int solution(String[] want, int[] number, String[] discount) {
        int answer = 0;
        Map<String, Integer> map = new HashMap<>();
        Queue<String> queue = new LinkedList<>();
        
        for (int i = 0; i < want.length; i++) {
            map.put(want[i], number[i]);
        }
        
        for (int i = 0; i < 10; i++) {
            queue.add(discount[i]);
        }

        for (int i = 10; i <= discount.length; i++) {
            Iterator<String> iter = queue.iterator();
            Map<String, Integer> chkMap = new HashMap<>();
            for (; iter.hasNext();) {
                String s = iter.next();
                if (chkMap.containsKey(s)) {
                    chkMap.put(s, chkMap.get(s) + 1);
                } else {
                    chkMap.put(s, 1);
                }
            }
            if (map.equals(chkMap)) {
                answer += 1;
            }
            if (i < discount.length){
            queue.remove();
            queue.add(discount[i]);
            }
        }
        
        return answer;
    }
}

File: CodingTest/Python/solution22.py
# https://school.programmers.co.kr/learn/courses/30/lessons/258712?language=python3
def solution(friends, gifts):
    answer = 0
    lst = [[0 for _ in range(len(friends))] for _ in range(len(friends))]
    
    # 선물 기록
    for i in gifts:
        A, B = i.split(" ")
        A = friends.index(A)
        B = friends.index(B)

        lst[A][B] += 1
    
    # 선물 지수
    lst2 = []
    for i in range(len(friends)):
        give = 0
        take = 0
        for j in range(len(friends)):
            if i != j:
                give += lst[i][j]
                take += lst[j][i]
        lst2.append(give - take)

    # 최대값 계산
    sol = [0 for i in range(len(friends))]
    for i in range(len(friends)):
        for j in range(len(friends)):
            if i == j:
                pass
            else:
                if lst[i][j] > lst[j][i]:
                    sol[i] += 1
                elif lst[i][j] == lst[j][i]:
                    if lst2[i] > lst2[j]:
                        sol[i] += 1
                    elif lst2[i] == lst2[j]:
                        pass
                    else:
                        pass
                else:
                    pass
    return max(sol)
